### PR TITLE
fixed undefined behaviour of sprintf, buffer overflow, crash

### PR DESCRIPTION
**Pull request by AMV007**

### CODE REVIEW

**Comment by bmiklautz** [comment]
@freerdp-bot test

**Comment by bmiklautz** [comment]
@AMV007 where exactly did you get the buffer overflow/crash?

**Comment by AMV007** [comment]
@bmiklautz 
exactly was buffer overflow at :
`sprintf(strContainerId,
		"{%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x}",
		containerId[0], containerId[1],containerId[2], containerId[3],
		containerId[4], containerId[5], containerId[6], containerId[7],
		containerId[8], containerId[9], containerId[10], containerId[11],
		containerId[12], containerId[13], containerId[14], containerId[15]);`

**Comment by nfedera** [comment]
@AMV007 IO this doesn't fix anything

**Comment by nfedera** [comment]
The real issue in func_container_id_generate() is the first sprintf creating the containerId:
```
sprintf(containerId, "%04X%04X%s", idVendor, idProduct, p);
```
And containerId, idVendor and idProduct are defined like this:
```
char containerId[17];
int idVendor, idProduct;
```
The code makes sure that p is not longer than 8. However, the original author seems to assume that the "%04X" format specifier would always generate 4 characters which is not true. 
So if idVendor/idProduct are negative or >= 2^16  th sprintf will write outside of containerId

File: channels/urbdrc/client/urbdrc_main.c [hunk 1]
```diff
@@ -50,9 +50,9 @@ static int func_hardware_id_format(IUDEVICE* pdev, char(*HardwareIds)[DEVICE_HAR
 	idProduct = pdev->query_device_descriptor(pdev, ID_PRODUCT);
 	bcdDevice = pdev->query_device_descriptor(pdev, BCD_DEVICE);
 
-	sprintf(str, "USB\\VID_%04X&PID_%04X", idVendor, idProduct);
+	sprintf(str, "USB\\VID_%04X&PID_%04X", (unsigned int)idVendor, (unsigned int)idProduct);
```
Comment (nfedera) [inline channels/urbdrc/client/urbdrc_main.c]: (unsigned int) does not fit into %04X !

File: channels/urbdrc/client/urbdrc_main.c [hunk 2]
```diff
@@ -152,15 +152,15 @@ static int func_container_id_generate(IUDEVICE* pdev, char* strContainerId)
 	else
 		p = path;
 
-	sprintf(containerId, "%04X%04X%s", idVendor, idProduct, p);
+	sprintf(containerId, "%04X%04X%s", (unsigned int)idVendor, (unsigned int)idProduct, p);
```
Comment (nfedera) [inline channels/urbdrc/client/urbdrc_main.c]: (unsigned int) does not fit into %04X

**Comment by AMV007** [comment]
ye, it's must be unsgined short, my mistake )

**Comment by AMV007** [comment]
should i close this pull and open new one ?

**Comment by AMV007** [comment]
fixed in patch

**Comment by nfedera** [comment]
@AMV007 that should do it. However, I'd personally prefer to use correct variables in the first place (e.g. UINT16 idVendor, idProduct;  or BYTE bDeviceSubClass)  then you don't need to cast a hundred times. And also use snprintf instead of sprintf !

**Comment by AMV007** [comment]
ye, better do work right, but have limited time usually, right now had task to quick fix freerdp crash )
if sprintf crashing - you can easely find crash place, but if snprintf will provide incorrect result-this bug may last years, if generally all will work ?

**Comment by nfedera** [comment]
@AMV007, no offense, but writing beyond valid buffers is NEVER better. For starters http://en.wikipedia.org/wiki/Buffer_overflow

**Comment by nfedera** [comment]
@AMV007 Nevertheless many thanks for helping and reporting the issue!
Don't worry because of your limited time. I've created a modified PR: #2531
Do you mind testing and reporting if if works for you ?

**Comment by AMV007** [comment]
@nfedera, no I haven't problem device in my hands, it was tester issue, and remote debug for find error, but your patch must work.
By the way you forgot couple sprintf change to snprintf,  look in this updated patch, I did some extend to your version.

**Comment by nfedera** [comment]
+1 LGTM

**Pull request merged**